docs: add event-bus, switchover and integrations

tiagolobocastro opened the pull request.

tiagolobocastro committed:
docs: add event-bus, switchover and integrations

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

Abhinandan-Purkait commented on doc/design/control-plane.md:
```diff
@@ -214,6 +214,56 @@ sequenceDiagram
+
+<br>
+
+As you can see, there's many p2p connections between the different services.
```
```suggestion
As we can see, there's many p2p connections between the different services.
```

Abhinandan-Purkait commented on doc/design/control-plane.md:
```diff
@@ -334,9 +384,8 @@ The value add is not the ANA feature itself, rather what you do with it.
-We used to use NATS as a message bus within mayastor as a whole, but as since switched for gRPC for p2p communications. \
-We will continue to use NATS for async notifications. Async in the sense that we send a message, but we do NOT wait for a reply. This mechanism does not
- do any form of "consensus," retries, and the likes. Information transported over NATS will typically be error telemetry that is used to diagnose problems. No work has started yet on this subject.
+We used to use [NATS] as a message bus within mayastor as a whole, but as since switched for [gRPC] for p2p communications. \
```
```suggestion
We used to use [NATS] as a message bus within Mayastor but later switched to [gRPC] for p2p communications. \
```

Abhinandan-Purkait commented on doc/design/control-plane.md:
```diff
@@ -334,9 +384,8 @@ The value add is not the ANA feature itself, rather what you do with it.
-We will continue to use NATS for async notifications. Async in the sense that we send a message, but we do NOT wait for a reply. This mechanism does not
- do any form of "consensus," retries, and the likes. Information transported over NATS will typically be error telemetry that is used to diagnose problems. No work has started yet on this subject.
+We used to use [NATS] as a message bus within mayastor as a whole, but as since switched for [gRPC] for p2p communications. \
+As of today, we use [NATS] for [events](./events.md) via async notifications. Async in the sense that we send a message, but we do NOT wait for a reply. This mechanism does not do any form of "consensus," retries, and the likes. Information transported over NATS will typically be error telemetry that is used to diagnose problems. No work has started yet on this subject.
```
Maybe can we reword or put somewhere we follow publisher subscriber model but don't have any active subscriber at the moment, instead of explaining in what sense it's async?

tiagolobocastro replied:
Reworded a bit

Abhinandan-Purkait commented on doc/design/control-plane.md:
````diff
@@ -458,23 +508,72 @@ An example where a `REST` client sets the open tracing key/values on the request
 })
 ```
 
+On the server-side we extra the trace id from the `HTTP` headers, and we inject it on the next call stack, which means it also gets eventually
````
```suggestion
On the server-side we extract the trace id from the `HTTP` headers, and we inject it on the next call stack, which means it also gets eventually
```

Abhinandan-Purkait commented on doc/design/control-plane.md:
````diff
@@ -458,23 +508,72 @@ An example where a `REST` client sets the open tracing key/values on the request
 ```
 
+On the server-side we extra the trace id from the `HTTP` headers, and we inject it on the next call stack, which means it also gets eventually
+injected in the next transport hoop. Specifically for `REST`, this means we inject it on the inter-service `gRPC`. Again, we use the same `Service`
````
```suggestion
injected in the next transport hop. Specifically for `REST`, this means we inject it on the inter-service `gRPC`. Again, we use the same `Service`
```

Abhinandan-Purkait commented on doc/design/events.md:
```diff
@@ -0,0 +1,155 @@
+| Nexus             | SubsystemResume   | Data plane | Created when an I/O subsystem is resumed             |
+| Nexus             | Init              | Data plane | Created when nexus enters into init state            |
+| Nexus             | Reconfiguring     | Data plane | Created when nexus enters into reconfiguring state   |
+| Nexus             | Shutdown          | Data plane | Created when a nexus is destroyed                    |
```
```suggestion
| Nexus             | Shutdown          | Data plane | Created when a nexus is shutdown                    |
```

Abhinandan-Purkait commented on doc/design/ha-failover.md:
```diff
@@ -0,0 +1,182 @@
+
+#### Nexus Shutdown
+
+In order to prevent applications from seeing reservation errors while failing over healthy nexuses (for instance, upon node draining), there should be a way to suspend I/O on nexus before reconstruct/re-publish the nexus on another node. To achieve this, a new "shutdown" operation shall be supported by I/O engine for nexuses, which shall do the following:
```
```suggestion
In order to prevent applications from seeing reservation errors while failing over healthy nexuses (for instance, upon node draining), there should be a way to suspend I/O on nexus before reconstructing/re-publish the nexus on another node. To achieve this, a new "shutdown" operation shall be supported by I/O engine for nexuses, which shall do the following:
```

Abhinandan-Purkait commented on doc/design/ha-node-agent.md:
```diff
@@ -0,0 +1,105 @@
+|------------|--------------------------------------------------------------|
+| live       | The Controller is connected and I/O capable                  |
+| connecting | The Controller is disconnected, now connecting the transport |
+| resetting  | The Controller is resetting (or scheduled reset)             |
```
what about `deleting(no io)` and `deleting`

Abhinandan-Purkait commented on doc/design/ha-node-agent.md:
```diff
@@ -0,0 +1,105 @@
+| resetting  | The Controller is resetting (or scheduled reset)             |
+
+Once a path failure is encountered at the transport level (`TCP` socket in case of `NVMe-oF`), a controller reset is started and the NVMe controller transitions into the "resetting" state.
+In this state, he controller only resets its core structures (like tearing-down I/O queues and stopping keep-alive timer), but doesn't reconnect I/O queues.
```
```suggestion
In this state, the controller only resets its core structures (like tearing-down I/O queues and stopping keep-alive timer), but doesn't reconnect I/O queues.
```

Abhinandan-Purkait commented on doc/design/ha-node-agent.md:
````diff
@@ -0,0 +1,105 @@
+    FAILED --> [*]
+```
+
+The HA node agent constantly monitors all suitable NVMe paths at `T/2` frequency to classify them as failed.
````
Is it T/2 or T?

tiagolobocastro replied:
It's T/2, so it reports it as T when it fails twice, IIRC

Abhinandan-Purkait commented on doc/design/obs-call-home.md:
```diff
@@ -0,0 +1,165 @@
+A component whose role is to store time series information for call-home and health metrics:
+
+- metrics which can’t be simply derived from current state of the control-plane or data-plane components
+  - i.e. number of replicas created since installation
```
Is indentation off here or it should just have been in the previous line?

tiagolobocastro replied:
It's intended this way

Abhinandan-Purkait commented on doc/integrations.md:
```diff
@@ -0,0 +1,37 @@
+| [NATS]                 | [Event bus]                                      | Used as event bus                                                     |
+| [OpenTelemetry]        | [Tracing]                                        | Tracing system for observability                                      |
+| [Helm]                 | [Helm Install Guide]                             | Installs/upgrades on K8s cluster                                      |
+| [Grafana]              | [Grafana Dashboards]                             | Install grafana custom dashboards with OpenEBS exported metrics       |
```
Missing loki?

tiagolobocastro replied:
We don't have any docs on Loki

Abhinandan-Purkait commented on doc/integrations.md:
```diff
@@ -0,0 +1,37 @@
+| [Prometheus]           | [Monitoring]                                     | Export stats                                                          |
+| [Kubernetes]           | [Install Guide]                                  | Runs on K8s                                                           |
+
+[//]: <>  (Technology Links)
+[Grafana]: https://grafana.com/
+[SPDK]: https://spdk.io/
+[gRPC]: https://grpc.io/
+[etcd]: https://etcd.io/
+[NATS]: https://nats.io/
+[OpenTelemetry]: https://opentelemetry.io/
+[Helm]: https://helm.sh/
+[Prometheus]: https://prometheus.io/
+[Kubernetes]: https://kubernetes.io/
+
+[//]: <>  (Integrations Links)
+[Grafana Dashboards]: https://openebs.io/docs/main/user-guides/observability#install-the-helm-chart
+[spdk-rs]: https://github.com/openebs/spdk-rs
+[io-engine]: https://github.com/openebs/mayastor/blob/HEAD/doc/design/mayastor.md
+[Inter Service Communication]: https://github.com/openebs/mayastor/blob/HEAD/doc/design/control-plane.md#internal-communication
+[Persistent Store]: https://github.com/openebs/mayastor/blob/HEAD/doc/design/control-plane.md#persistent-store-kvstore-for-configuration-data
+[PStor client]: https://github.com/openebs/mayastor-control-plane/blob/HEAD/utils/pstor/src/etcd.rs
+[Event bus]: https://github.com/openebs/mayastor/blob/HEAD/doc/design/events.md
+[Tracing]: https://github.com/openebs/mayastor/blob/HEAD/doc/design/control-plane.md#tracing-and-telemetry
+[Helm Install Guide]: https://openebs.io/docs/quickstart-guide/installation#installation-via-helm
+[Monitoring]: https://openebs.io/docs/user-guides/replicated-storage-user-guide/replicated-pv-mayastor/advanced-operations/monitoring
+[Install Guide]: https://openebs.io/docs/quickstart-guide/installation
```
This is not visible in rich text for some reason? I only see the above table

tiagolobocastro replied:
Right, so it's working :)

avishnu commented on doc/design/control-plane.md:
```diff
@@ -334,9 +384,10 @@ The value add is not the ANA feature itself, rather what you do with it.
-We used to use NATS as a message bus within mayastor as a whole, but as since switched for gRPC for p2p communications. \
-We will continue to use NATS for async notifications. Async in the sense that we send a message, but we do NOT wait for a reply. This mechanism does not
- do any form of "consensus," retries, and the likes. Information transported over NATS will typically be error telemetry that is used to diagnose problems. No work has started yet on this subject.
+We used to use [NATS] as a message bus within mayastor as a whole, but later switched to [gRPC] for p2p communications. \
```
We can remove this IMO.

avishnu commented on doc/design/control-plane.md:
```diff
@@ -334,9 +384,10 @@ The value add is not the ANA feature itself, rather what you do with it.
+We used to use [NATS] as a message bus within mayastor as a whole, but later switched to [gRPC] for p2p communications. \
+As of today, we use [NATS] for [events](./events.md) as async notifications via the [Publish/Subscribe Model](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern). Our current mechanism does not do any form of "consensus," retries, and the likes. Information transported over NATS will typically be error telemetry that is used to diagnose problems.
+
+> NOTE: We don't have any subscribers as things stand.
```
The cal-home stats aggregator is a subscriber IIRC.

tiagolobocastro replied:
Indeed, done

avishnu commented on doc/design/control-plane.md:
```diff
@@ -411,21 +462,22 @@ message EventMessage {
-forwarded to, Elastic Search, Cassandra, Kafka, or whatever. In order to achieve full tracing support, all the gRPC requests and replies should add
-HTTP headers such that we can easily tie them together in whatever tooling is used. This is standard practice but requires a significant amount of work.
-The key reason is to ensure that all requests and responses pass along the headers, from REST to the scheduling pipeline.
+Tracing is by default implemented using open telemetry and, by default, we have provided a subscriber for [Jaeger]. From [Jaeger], the information can be
```
Nit: 'by default' is repeated

avishnu commented on doc/design/control-plane.md:
```diff
@@ -411,21 +462,22 @@ message EventMessage {
+We also need to support several types of transport and serialization mechanisms. For example, `HTTP/1.1 REST` requests to `HTTP/2 gRPC` request to
  a KV store operation to etcd. For this, we will use [Tower]. \
-[Tower] provides a not-so-easy to use an abstraction of Request to Response mapping.
+[Tower] provides a not-so-easy to use an abstraction of request to response mapping.
```
Nit: Sentence seems incomplete, 'not-so-easy to use...'

Maybe, you mean, not-so-easy framework or interface or tooling?

tiagolobocastro replied:
Done

avishnu commented on doc/design/events.md:
```diff
@@ -0,0 +1,155 @@
+# Event Bus
+
+As part of mayastor we wanted to have event driven capabilities which can allow us to respond to certain events and perform specific actions. \
+A message bus ([NATS]) was initially used in early versions of mayastor, but due to several reasons such as some bugs in the client libraries and how we used them (to be fair!), we ended up temporarily moving away from it in favour of p2p [gRPC] (yes we were probably using the wrong stick if many cases). \
```
Again, I think this may not be needed.

tiagolobocastro replied:
Yeah, I've reworded it a bit

avishnu commented on doc/design/events.md:
```diff
@@ -0,0 +1,155 @@
+
+---
+
+We don't currently have a requirement for a messaging queue where order is important, nor do we rely or this information to be persistent. \
```
rely on* this information

avishnu commented on doc/design/obs-call-home.md:
```diff
@@ -0,0 +1,165 @@
+
+A call-home deployment, the "transmission module", periodically sends reports about the installation, to a remote consumer. It obtains this information via:
+
+- "state metrics": periodic queries to the `REST API` to determine the current state
```
stats instead of state?

avishnu replied:
Ignore the above comment.

avishnu reviewed (approved):
Few minor suggestions. LGTM.

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
Something wrong with the runners ??
```text
 Ign:11 http://security.ubuntu.com/ubuntu noble-updates/main amd64 libc6-dbg amd64 2.39-0ubuntu8.3
Err:11 mirror+file:/etc/apt/apt-mirrors.txt noble-updates/main amd64 libc6-dbg amd64 2.39-0ubuntu8.3
  404  Not Found [IP: 52.250.76.244 80]
E: Failed to fetch mirror+file:/etc/apt/apt-mirrors.txt/pool/main/g/glibc/libc6-dbg_2.39-0ubuntu8.3_amd64.deb  404  Not Found [IP: 52.250.76.244 80]
Fetched 4891 kB in 2s (2459 kB/s)
```

bors merge

tiagolobocastro commented:
bors merge

tiagolobocastro committed:
ci(gha): coredump before cleanup

List all coredumps since start of testing.
Ensure unlimited coredump.
Run coredump check before cleaning the workspace. This is helpful in
case there was a coredump from our own binaries.
Also disable the man-pages auto-update.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
bors merge
